source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.cc

```diff
@@ -79,14 +79,29 @@ void ShadowRouterImpl::flushPendingCallbacks() {
   pending_callbacks_.clear();
 }
 
+FilterStatus ShadowRouterImpl::runOrSave(std::function<FilterStatus()>&& cb,
+                                         const std::function<void()>& on_save) {
+  if (requestStarted()) {
+    return cb();
+  }
+
+  pending_callbacks_.push_back(std::move(cb));
+
+  if (on_save) {
+    on_save();
+  }
+
+  return FilterStatus::Continue;
+}
+
 FilterStatus ShadowRouterImpl::passthroughData(Buffer::Instance& data) {
   if (requestStarted()) {
     return ProtocolConverter::passthroughData(data);
   }
 
   auto copied = std::make_shared<Buffer::OwnedImpl>(data);
   auto cb = [copied = std::move(copied), this]() mutable {
-    ProtocolConverter::passthroughData(*copied);
+    return ProtocolConverter::passthroughData(*copied);
```

Comment:
Add `-> FilterStatus` to the lambda definition (similar for the ones in structBegin and structEnd below)?